core/consensus/pow/src/lib.rs

```diff
@@ -0,0 +1,301 @@
+		{
+			Some(bytes) => PowAux::decode(&mut &bytes[..]).map_err(|e| format!("{:?}", e))?,
+			None => Default::default(),
+		};
+
+		let (checked_header, difficulty, seal) = self.check_header::<B>(
+			header,
+			BlockId::Hash(parent_hash),
+		)?;
+		aux.total_difficulty += difficulty;
+
+		if let Some(inner_body) = body.take() {
+			let block = B::new(checked_header.clone(), inner_body);
+
+			self.check_inherents(
+				block.clone(),
+				BlockId::Hash(parent_hash),
+				inherent_data,
+				timestamp_now
+			)?;
+
+			let (_, inner_body) = block.deconstruct();
+			body = Some(inner_body);
+		}
+		let key = POW_AUX_PREFIX.iter().chain(&hash[..]).cloned().collect::<Vec<_>>();
+		let import_block = BlockImportParams {
+			origin,
+			header: checked_header,
+			post_digests: vec![seal],
+			body,
+			finalized: false,
+			justification,
+			auxiliary: vec![(key, Some(aux.encode()))],
+			fork_choice: ForkChoiceStrategy::Custom(aux.total_difficulty > best_aux.total_difficulty),
+		};
+
+		Ok((import_block, None))
+	}
+}
+
+/// The PoW import queue type.
+pub type PowImportQueue<B> = BasicQueue<B>;
+
+pub fn import_queue<B, C>(
+	block_import: BoxBlockImport<B>,
+	client: Arc<C>,
+	inherent_data_providers: InherentDataProviders,
+) -> Result<PowImportQueue<B>, consensus_common::Error> where
+	B: BlockT<Hash=H256>,
+	C: 'static + ProvideRuntimeApi + HeaderBackend<B> + BlockOf + ProvideCache<B> + Send + Sync + AuxStore,
+	C::Api: BlockBuilderApi<B> + PowApi<B>,
+{
+	if !inherent_data_providers.has_provider(&srml_timestamp::INHERENT_IDENTIFIER) {
+		match inherent_data_providers.register_provider(srml_timestamp::InherentDataProvider) {
+			Ok(()) => (),
+			Err(e) => warn!("Registering inherent data provider for timestamp failed"),
+		}
+	}
+
+	let verifier = PowVerifier {
+		client: client.clone(),
+		inherent_data_providers,
+	};
+
+	Ok(BasicQueue::new(
+		verifier,
+		block_import,
+		None,
+		None
+	))
+}
+
+pub fn start_mine<B: BlockT<Hash=H256>, I, C, E>(
+	block_import: Arc<Mutex<I>>,
+	client: Arc<C>,
+	mut env: E,
+	inherent_data_providers: inherents::InherentDataProviders,
+) where
+	I: BlockImport<B> + Send + Sync + 'static,
+	C: HeaderBackend<B> + AuxStore + ProvideRuntimeApi + 'static,
+	C::Api: PowApi<B>,
+	E: Environment<B> + Send + Sync + 'static,
+	E::Error: core::fmt::Debug,
+{
+	if !inherent_data_providers.has_provider(&srml_timestamp::INHERENT_IDENTIFIER) {
+		match inherent_data_providers.register_provider(srml_timestamp::InherentDataProvider) {
+			Ok(()) => (),
+			Err(e) => warn!("Registering inherent data provider for timestamp failed"),
+		}
+	}
+
+	thread::spawn(move || {
+		loop {
+			match mine_one(&block_import, &client, &mut env, &inherent_data_providers) {
+				Ok(()) => (),
+				Err(e) => warn!("Mining block failed: {:?}", e),
+			}
+
+			std::thread::sleep(std::time::Duration::new(1, 0));
+		}
+	});
+}
+
+pub fn mine_one<B: BlockT<Hash=H256>, I, C, E>(
+	block_import: &Arc<Mutex<I>>,
```

Comment:
Related to the threading question: I'm reaaally not a fan of having `Mutex`es, and I'm generally trying to remove `Mutexes` from the codebase entirely (cc #2536).

Wouldn't it be preferable for `mine_one` to return some sort of "sealed block" object, and do the importing somewhere else?
I'm also generally trying to un-spaghettify the general design of Substrate, and passing the import queue by parameter to a background thread is really not great IMO.

Reply:
I figured out that this `Mutex` is not necessary at all! We can use `BoxBlockImport` because `Client` already handles locking for its `BlockImport` implementation.